File: app/src/main/java/com/example/chuckjokes/BrowserFragment.java
package com.example.chuckjokes;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.WebChromeClient;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;

import androidx.fragment.app.Fragment;

import com.example.chuckjokes.databinding.BrowserFragmentBinding;

import org.jetbrains.annotations.NotNull;

public class BrowserFragment extends Fragment {

    public static final boolean LOAD_WITH_OVERVIEW_MODE = true;
    public static final boolean USE_WIDE_VIEWPORT = true;
    public static final boolean BUILT_IN_ZOOM_CONTROLS = true;
    public static final boolean DISPLAY_ZOOM_CONTROLS = false;
    public static final boolean JAVA_SCRIPT_ENABLED = true;
    public static final int MAX_PROGRESS = 100;

    public static final String DEFAULT_URL = "http://www.icndb.com/api/";

    private BrowserFragmentBinding binding;

    @Override
    public View onCreateView(@NotNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {

        binding = BrowserFragmentBinding.inflate(inflater);

        binding.browser.setWebViewClient(new WebViewClient() {
            @Override
            public boolean shouldOverrideUrlLoading(WebView view, String url) {
                view.loadUrl(url);
                return true;
            }
        });

        binding.browser.setWebChromeClient(new WebChromeClient() {
            @Override
            public void onProgressChanged(WebView wv, int progress) {
                binding.progressBar.setProgress(progress);
                if (progress == MAX_PROGRESS) {
                    binding.progressBar.setVisibility(View.GONE);
                } else {
                    binding.progressBar.setVisibility(View.VISIBLE);
                }
            }
        });

        WebSettings settings = binding.browser.getSettings();
        settings.setLoadWithOverviewMode(LOAD_WITH_OVERVIEW_MODE);
        settings.setUseWideViewPort(USE_WIDE_VIEWPORT);
        settings.setBuiltInZoomControls(BUILT_IN_ZOOM_CONTROLS);
        settings.setDisplayZoomControls(DISPLAY_ZOOM_CONTROLS);
        settings.setJavaScriptEnabled(JAVA_SCRIPT_ENABLED);

        binding.browser.loadUrl(DEFAULT_URL);

        return binding.getRoot();
    }
}

File: app/src/main/java/com/example/chuckjokes/fragments/browser/BrowserFragment.kt
package com.example.chuckjokes.fragments.browser

import android.annotation.SuppressLint
import android.view.LayoutInflater
import android.view.ViewGroup
import android.os.Bundle
import android.view.View
import android.webkit.WebChromeClient
import android.webkit.WebView
import android.webkit.WebViewClient
import androidx.fragment.app.Fragment
import com.example.chuckjokes.databinding.BrowserFragmentBinding
import com.example.chuckjokes.util.viewModelsExt

const val LOAD_WITH_OVERVIEW_MODE = true
const val USE_WIDE_VIEWPORT = true
const val BUILT_IN_ZOOM_CONTROLS = true
const val DISPLAY_ZOOM_CONTROLS = false
const val JAVA_SCRIPT_ENABLED = true

class BrowserFragment : Fragment() {

    private var _binding: BrowserFragmentBinding? = null
    private val binding
        get() = _binding!!

    private val viewModel by viewModelsExt {
        BrowserViewModel()
    }

    @SuppressLint("SetJavaScriptEnabled")
    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = BrowserFragmentBinding.inflate(inflater)

        savedInstanceState?.apply {
            //TODO: UI не сохраняет, но хотяб не перезагружается при перевороте, зато слетает масштабирование загруженной страницы
            binding.browser.restoreState(this)
        } ?: run {
            binding.browser.webViewClient = object : WebViewClient() {
                override fun shouldOverrideUrlLoading(view: WebView, url: String): Boolean =
                    viewModel.shouldOverrideUrlLoading()
            }

            binding.browser.webChromeClient = object : WebChromeClient() {
                override fun onProgressChanged(wv: WebView, progress: Int) {
                    viewModel.changeProgress(progress)
                }
            }

            binding.browser.settings.apply {
                loadWithOverviewMode = LOAD_WITH_OVERVIEW_MODE
                useWideViewPort = USE_WIDE_VIEWPORT
                builtInZoomControls = BUILT_IN_ZOOM_CONTROLS
                displayZoomControls = DISPLAY_ZOOM_CONTROLS
                javaScriptEnabled = JAVA_SCRIPT_ENABLED
            }
        }

        viewModel.progress.observe(viewLifecycleOwner) {
            binding.progressBar.progress = it
        }

        viewModel.visibility.observe(viewLifecycleOwner) {
            binding.progressBar.visibility = if (it) {
                View.VISIBLE
            } else {
                View.GONE
            }
        }

        binding.browser.saveState(Bundle())

        viewModel.url.observe(viewLifecycleOwner) {
            binding.browser.loadUrl(it)
        }

        return binding.root
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }
}